smcontroller/smctrlserver.go

```diff
@@ -0,0 +1,190 @@
+		controllerCh: controllerCh, runStatusCh: runStatCh, updateInstanceStatusCh: updateInstanceStatusCh,
+	}
+
+	var opts []grpc.ServerOption
+
+	if !insecure {
+		certURL, keyURL, err := certProvider.GetCertificate(cfg.CertStorage, nil, "")
+		if err != nil {
+			return nil, aoserrors.Wrap(err)
+		}
+
+		tlsConfig, err := cryptcoxontext.GetClientMutualTLSConfig(certURL, keyURL)
+		if err != nil {
+			return nil, aoserrors.Wrap(err)
+		}
+
+		opts = append(opts, grpc.Creds(credentials.NewTLS(tlsConfig)))
+	} else {
+		log.Info("GRPC server starts in insecure mode")
+	}
+
+	server.url = cfg.SMController.SMServerURL
+
+	server.grpcServer = grpc.NewServer(opts...)
+
+	pb.RegisterSMServiceServer(server.grpcServer, server)
+
+	return server, nil
+}
+
+// Start starts update controller server.
+func (server *smCtrlServer) Start() (err error) {
+	server.listener, err = net.Listen("tcp", server.url)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	go func() {
+		if err := server.grpcServer.Serve(server.listener); err != nil {
+			log.Errorf("Can't serve gRPC server: %s", err)
+		}
+	}()
+
+	return nil
+}
+
+// Stop stops update controller server.
+func (server *smCtrlServer) Stop() {
+	if server.grpcServer != nil {
+		server.grpcServer.Stop()
+	}
+
+	if server.listener != nil {
+		server.listener.Close()
+	}
+}
+
+// RegisterSM registers new SM client connection.
+func (server *smCtrlServer) RegisterSM(stream pb.SMService_RegisterSMServer) error {
+	message, err := stream.Recv()
+	if errors.Is(err, io.EOF) {
+		log.Warn("Unexpected end of UM stream")
+
+		return nil
+	}
+
+	if err != nil {
+		log.Error("Error receive message from SM ", err)
+
+		return aoserrors.Wrap(err)
+	}
+
+	nodeConfig, ok := message.SMOutgoingMessage.(*pb.SMOutgoingMessages_NodeConfiguration)
+	if !ok {
+		log.Error("unexpected first messager from stream")
+
+		return aoserrors.New("Unexpected first messager from stream")
+	}
+
+	log.Debugf("Register SM id %s", nodeConfig.NodeConfiguration.NodeId)
+
+	nodeCfg := launcher.NodeConfiguration{
+		NodeInfo: cloudprotocol.NodeInfo{
+			NodeID: nodeConfig.NodeConfiguration.NodeId, NumCPUs: nodeConfig.NodeConfiguration.NumCpus,
+			TotalRAM:   nodeConfig.NodeConfiguration.TotalRam,
+			Partitions: make([]cloudprotocol.PartitionInfo, len(nodeConfig.NodeConfiguration.Partitions)),
+		},
+		RemoteNode:    nodeConfig.NodeConfiguration.RemoteNode,
+		RunnerFeature: message.GetNodeConfiguration().RunnerFeatures,
+	}
+
+	for i, pbPartition := range nodeConfig.NodeConfiguration.Partitions {
+		nodeCfg.Partitions[i] = cloudprotocol.PartitionInfo{
+			Name: pbPartition.Name,
+			Type: pbPartition.Type, TotalSize: pbPartition.TotalSize,
+		}
+	}
+
+	handler, err := newSmHandler(stream, server.messageSender, server.alertSender, server.monitoringSender, nodeCfg,
+		server.runStatusCh, server.updateInstanceStatusCh)
+	if err != nil {
+		return err
+	}
+
+	openConnectionMsg := smCtrlInternalMsg{
```

Comment:
Why do we need separate struct for SM server? It has only one function. It can be part of SM controller. And sending private messages complicates implementation.